### PR TITLE
Pull Req: Added --params option to pass additional params to downstream tools

### PR DESCRIPTION
**Pull request by rwmtse**
User Story: blackberry/BB10-Webworks-Packager#157
CI Build: http://ci0000003863287:8080/hudson/job/BB10-Webworks-Packager-next-params-tools/5/
Microsite Issue/Sample Usage: http://github.rim.net/Microsites/HTML5-WebWorks/issues/193

@issp

### CODE REVIEW

**Comment by rwmtse** [comment]
@adrianlee Can you code review?

File: lib/signing-helper.js [hunk 1]
```diff
@@ -48,6 +53,50 @@ function getDefaultPath(file) {
...
+    args.push("-keystore");
+    args.push(session.keystore);
+    args.push("-storepass");
+    args.push(session.storepass);
+
+    if (params) {
+        Object.getOwnPropertyNames(params).forEach(function (p) {
+            args.push(p);
+
+            if (params[p]) {
```
Comment (adrianlee) [inline lib/signing-helper.js:76]: What happens if the param value is false. i don't know if there will be ever a case like that for the signer though

Comment (rwmtse) [inline lib/signing-helper.js:76]: If params is null/undefined, this means no additional params will be used for blackberry-signer.

**Comment by adrianlee** [comment]
Okay, everything looks good to me but there seems to be missing a test case for the nativepackager params.

**Comment by rwmtse** [comment]
Right, let me add one for native packager

**Comment by adrianlee** [comment]
r+ code review complete

**Comment by rwmtse** [comment]
@issp Please use this build for your testing:
http://ci0000003863287:8080/hudson/job/BB10-Webworks-Packager-next-params-tools/7/

**Comment by jengee** [comment]
Feature testing completed. All testcases passed.

**Comment by ishneur** [comment]
Tested on Windows and Mac.

There are a couple of issues that I noticed. Not sure whether they are considered bugs or not:
1. If in my .json file I specify params for blackberry-nativepackager and tell it to install the app, during packaging both the simulator bar and the device bar are deployed, with the simulator bar being deployed first and then the device bar overwriting it. Not sure whether there is a way to prevent it in the json file. This is the output:
   [INFO]    Populating application source
   [INFO]    Parsing config.xml
   [INFO]    Generating output files
   [INFO]    Info: Package created: C:\WebWorks\Widgets\BB10\API_Testing\simulator\
   APISmokeTest.bar
   [INFO]    Info: Sending request: Install
   [INFO]    Info: Resetting stale session
   [INFO]    Info: Sending request: Install
   [INFO]    Info: Action: Install
   [INFO]    Info: File size: 1150376
   [INFO]    Info: Installing ...Info: Processing 1150376 bytes
   [INFO]    Info: Progress 7%...
   [INFO]    Info: Progress 100%...
   [INFO]    Info: Progress 100%...
   [INFO]    actual_dname::APISmokeTest.testRel_PISmokeTest1db56dc7
   [INFO]    actual_id::testRel_PISmokeTest1db56dc7actual_version::1.2.3.16result::
   success
   [INFO]    Info: Package created: C:\WebWorks\Widgets\BB10\API_Testing\device\API
   SmokeTest.bar
   [INFO]    Info: Sending request: Install
   [INFO]    Info: Action: Install
   [INFO]    Info: File size: 1137320
   [INFO]    Info: Installing ...Info: Processing 1137320 bytes
   [INFO]    Info: Progress 40%...
   [INFO]    Info: Progress 100%...
   [INFO]    Info: Progress 100%...
   [INFO]    actual_dname::APISmokeTest.testRel_PISmokeTest1db56dc7actual_id::testR
   el_PISmokeTest1db56dc7
   [INFO]    actual_version::1.2.3.16result::success
   [INFO]    Info: Bar signed.
   [INFO]    BAR packaging complete
2. If in my json file I call both blackberry-signer and blackberry-nativepackager to sign and deploy, it looks like the bar is deployed before signing, so the bar deployed to the device is unsigned. It can also be seen in the above output.

The final bar on my desktop contains "Package-Author-Certificate-Hash: mud1WyQkseGkPFj3LeLWKETe0oxPTZNoni1Aa86Im4Nc1rqo8uDwpRWlGI4A4T3rr9epSKYlqGc76hp3h94fcA" in the manifest file, but the bar file on the device doesn't have it in the manifest (I checked it via ssh since my device is not secure). The final bar also contains AUTHOR and RDK files, but I don't see those on the device inside the bar.

The json file I used is:
{
    "_comment": "You can specify additional parameters to be passed to downstream tools by using the bbwp --params option",
    "blackberry-signer": {
        "-proxyhost": "wps01-vmyyz.labyyz.testnet.rim.net",
        "-proxyport": "80",
        "-proxyusername": "user01",
        "-proxypassword": "password"
    },
    "blackberry-nativepackager": {
        "-installApp": "",
        "-device": "10.137.234.255",
        "-password": "qaqa"
    }
}

The signing test was more relevant on Mac than on Windows, since even if Windows is configured to use a proxy, signing still works without specifying parameters.

**Comment by rwmtse** [comment]
@kjwallis @nukulb see Igor's test results. If we want this feature to work for more cases, we might need to take into consideration for more things, like allowing params to be specified for simulator/device build etc.

----- Original Message -----

**Comment by jengee** [comment]
Nice catch. I didn't think of using it in that way. I guess it is always good to have more than 1 person to look at it.

# -j

Sent from a Blackberry device while not driving.

----- Original Message -----

**Comment by kwallis** [comment]
We can leave this feature as is.  The goal was a simple pass through.  If we want to handle the actual use case of installing the app with bbwp it should be done in a different way.

**Comment by nukulb** [comment]
@rwmtse - rebase please

**Comment by nukulb** [comment]
@ishneur / @issp - is this tested fully now?
Ready for merge?

**Comment by ishneur** [comment]
Retested this on Windows using build http://ci0000003863287:8080/hudson/job/BB10-Webworks-Packager-next-params-tools/8/ and no new issues found.

I already tested it before the rebase and posted my comments above. The same issues still apply.

**Comment by ishneur** [comment]
If I specify a params file which doesn't exist, then packaging fails but the packager still creates an "src" and a "simulator" folder in the directory in which the zip file is and doesn't clean them up.

This is a packager problem not directly related to this feature, so we can just open a separate issue for this. It could be related to blackberry/BB10-Webworks-Packager#51.

**Comment by tohman** [comment]
@rwmtse - can you please rebase into one commit so that we can merge easily? right now we are getting merge conflicts

**Comment by rwmtse** [comment]
@tohman @nukulb Rebased, squashed, smoked tested

**Comment by nukulb** [comment]
merged manually

**Pull request merged**